packages/gatsby/src/db/sanitize-node.js

```diff
@@ -10,13 +10,12 @@ const sanitizeNode = (data, isNode = true, path = new Set()) => {
   const isPlainObject = _.isPlainObject(data)
 
   if (isPlainObject || _.isArray(data)) {
+    if (path.has(data)) return data
     path.add(data)
```

Comment:
This is a bit problematic because we short circuit to use input objects and those are unsanitized.

So if in our tests we use something like this as input:
```diff
    const circularReference = {
+      fn: () => {},
    }
    circularReference.self = circularReference
```

We will get snapshot like this:
```diff
Object {
    "children": Array [],
    "circularReference": Object {
-     "self": [Circular],
+     "self": Object {
+       "fn": [Function],
+       "self": [Circular],
+     },
    },
```

This is problematic, because to fix that we would need to use something like WeakMap with input objects as keys and sanitized objects as values. The problem is that we get sanitized values only after full traversal, so we would still have infinite recursion. So this becomes pretty complex to solve this properly I think?

so "first circular reference" get sanitized and `fn` is removed, but repeated nested objects are not sanitized

Reply:
TIL about ` ```diff ` :D

@pieh Will discuss this in chat, I don't quite see the problem you're describing just yet.